More stock assistant

jralls opened the pull request:
Fixes the bugs and issues, almost finishes the reorganization, adds Doxygen documentation.

jralls committed:
Make StockAssistantController's view member in-place instead of a unique_ptr.

Requires moving some parameters around so we can initialize it
correctly:
All of the view's pages need to take an Account* and get its commodity
or currency in their ctors, and creation of the builder has to move to
gnc_stock_transaction_assistant and get passed in.

jralls committed:
Make connect_signals and prepare StockAssistantController members.

jralls committed:
[stock-txn-asst] Move prepare details into Assistant Page member funcs.

Substantially simplifies the code by reducing the needed indirection.
In most cases the StockTransactionEntry has everything the page needs
and can be passed as a parameter, further simplifying the code.

jralls committed:
[stock-txn-asst] Extract Logger class

jralls committed:
[stock-txn-asst] Fix sell-side stock amounts.

It was setting a negative amount in one too many places.

jralls committed:
[asst-stock-txn] Fix typo.

jralls committed:
Bug 799054 - Stock Assist not functioning

* Get the sign right when calculating new shares
* Fix displaying values and units on the summary page
* Don't display "missing" for values when the user skips
  entering a 0 value.
* Create the Stock-account's fees split (when capitalizing)
  and cap-gains split.

jralls committed:
[stock-txn-asst] Don't generate stock splitinfo if no stock split.

jralls committed:
[stock-txn-asst] Call gettext on the transaction type explanation.

jralls committed:
[stock-txn-asst] Add notes about cash-in-lieu and capgains.

Cash in lieu must be reported as a separate transaction. There was
already a note about that in reverse splits, apply it to regular
splits.

Add suggestion to enter a placeholder number when one can't
immediately calculate the capital gain on a sale or cover buy.

jralls committed:
[stock-txn-asst] Reformat explanation messages to remove long lines.

jralls committed:
[stock-txn-asst] Make action a StockTransactionEntry member.

Instead of a function paramter that's set and translated
inconsistently.

jralls committed:
[stock-txn-asst] Remove StockTransactionSplitInfo

With action a StockTransactionEntry member it's no longer useful.

jralls committed:
[stock-txn-asst] Move stock amount calculation to StockEntry.

It's StockEntry's responsibility not Model's. This lets us remove
the model parameter from PageStockAmount::prepare. It also requires
that we use StockEntry's set_amount method instead of just setting
m_amount it StockAssistantTest::instantiate_model.

jralls committed:
[stock-txn-asst] Set all the needed parameters for StockCGEntry.

So that it will create the split and do so with the right memo.

jralls committed:
[stock-txn-asst] Handle correctly capitalizing fees.

jralls committed:
[stock-txn-asst] Fix setting the initial focussed widget.

One can't grab the widget in prepare, the page runs its own focus so
one must handle the page's focus event.

jralls committed:
[stock-txn-asst] Separate setting the amount/value and validating it.

Setting happens via callbacks and they were handed direct access
to the member variables so the validation wasn't happening.
Validation also requires a logger and passing that to the callbacks
would be hard.

jralls committed:
[stock-txn-asst] model::calculate_price becomes entry::print_price

Replace multiple-responsibility function in the wrong class with a
single-responsibility function in the right class.

Changes the signature of a couple of PageStockValue functions.

jralls committed:
[stock-txn-asst] Variable-specific callbacks take an entry pointer.

Prerequisite for making class member variables private.

jralls committed:
[stock-txn-asst] Change structs to classes.

With private data members, replacing direct member access with
accessor functions and refactoring functionality away from the
Model, View, and Controller container classes to the classes
that own the members.

This removes the need for RTTI inspection as regular virtual
function dispatch accounts for differences in Entry classes.

jralls committed:
[stock-txn-asst] Ensure correct memory management flow.

For the StockTransferController and GncGuiManager events.

jralls committed:
[stock-txn-asst] Doxygen documentation of classes.

And non-trivial functions.

christopherlam commented:
not running well yet...

jralls commented:
@christopherlam 
>not running well yet...

Can you be more specific, starting with what OS?

christopherlam commented:
Linux as usual

From stable: `git pull --no-rebase https://github.com/jralls/gnucash.git more-stock-assistant` and recompile, no issues. Then:

```
* 12:06:01 ERROR <gnc.gui> void StockAssistantController::prepare(GtkAssistant*, GtkWidget*): assertion 'm_model->txn_type_valid()' failed
```

jralls commented:
@christopherlam Got it, thanks. Should be fixed now.

christopherlam commented:
Much better. I played a bit, there _should_ be an error if eg. a dividend type transaction, and the dividend account is missing. Lots of the following:

```
* 19:40:57 ERROR <gnc.engine> gnc_commodity* gnc_account_get_currency_or_parent(const Account*): assertion 'account' failed
* 19:41:55 ERROR <gnc.engine> gnc_commodity* gnc_account_get_currency_or_parent(const Account*): assertion 'account' failed
* 19:42:39 ERROR <gnc.engine> gnc_commodity* gnc_account_get_currency_or_parent(const Account*): assertion 'account' failed
```

jralls committed:
[stock-txn-asst] Remove unused method.

jralls commented:
OK, I added an error for there not being an account when there should be, to display the stock account on the fees line when capitalizing fees, and an if to not try to print the amount if there's no account to get the currency from.

The macOS test is failing because of a Sourceforge problem. We can ignore that for now.

christopherlam commented:
ok the GNCAccountSels don't seem to work now.

christopherlam commented:
> ok the GNCAccountSels don't seem to work now.

Ok further clarification: I used 797796's datafile. I tried to recreate the dividend transaction. I couldn't, because 797796.gnucash's SPY's parent is in CAD and the assistant will set the txn_currency to be CAD. Conclusion: given the assitant cannot create a txn whose currency isn't the stock's parent currency, I think this is acceptable.

Here's a dividend issue: creating a simple dividend should add a zero-amt, zero-value split in the stock acct, and it doesn't.

christopherlam commented:
Ok now this branch would want `gnc_account_sel_set_default_new_commodity (GNC_ACCOUNT_SEL(m_selector), currency);` somewhere

jralls committed:
[stokc-txn-asst] Add default currency to GncAccountSelector

Set the default currency to the transaction currency for creating
new accounts.

jralls committed:
[stock-txn-asst] Include a marker split in the stock account for divs.

jralls committed:
[stock-txn-asst] Add multicurrency note to introduction.

Warning the user that they're not supported.

jralls committed:
[stock-txn-asst] Don't report account missing if it's not required.

jralls commented:
@christopherlam Added the new account default currency, the tracking split for dividends, the multicurrency note to the intro, and fixed showing a missing account for fees when the amount was GNC_ERROR_ARG.

christopherlam commented:
It seems to work well so far.

A small comment: this being a stock assistant, it seems right to assume that *all* txn_types _will_ insert a stock-acct split. Therefore in my design I didn't specify a separate `MARKER_SPLIT` flag.

Moreover, `FieldMask::DISABLED` was intended to fully denote fields which should skip the page i.e. in my original design, the dividend page should skip both stock_amt and stock_val. In this branch, the "Stock Split" type will define stock_amt as `FieldMask::DISABLED | FieldMask::AMOUNT_DEBIT | FieldMask::INPUT_NEW_BALANCE` which is identical to `FieldMask::AMOUNT_DEBIT | FieldMask::INPUT_NEW_BALANCE`, i.e. it seems odd.

christopherlam commented on gnucash/gnome/assistant-stock-transaction.cpp:
```diff
@@ -292,7 +261,10 @@ reinvested must be subsequently recorded as a regular stock purchase.")
-        N_("Company issues additional units, thereby reducing the stock price by a divisor, while keeping the total monetary value of the overall investment constant.")
+        N_("Company issues additional units, thereby reducing the stock price by a divisor "
+           ", while keeping the total monetary value of the overall investment constant. "
+           "\n\nIf the reverse split results in a cash in lieu for remainder units, please "
```
```suggestion
           "\n\nIf the split results in a cash in lieu for remainder units, please "
```

christopherlam commented on gnucash/gnome/assistant-stock-transaction.cpp:
```diff
@@ -401,58 +387,221 @@ static const TxnTypeVec short_types
-        N_("Company issues additional units, thereby reducing the stock price by a divisor, while keeping the total monetary value of the overall investment constant.")
+        N_("Company issues additional units, thereby reducing the stock price by a divisor, "
+           "while keeping the total monetary value of the overall investment constant. "
+           "\n\nIf the reverse split results in a cash in lieu for remainder units, please "
```
```suggestion
           "\n\nIf the split results in a cash in lieu for remainder units, please "
```

jralls replied:
Good catch, thanks.

christopherlam reviewed (approved):
from my testing today everything seems ok. comments about marker_split and FieldMask::DISABLED remains. minor explanation fix.

jralls commented:
> A small comment: this being a stock assistant, it seems right to assume that all txn_types will insert a stock-acct split. Therefore in my design I didn't specify a separate MARKER_SPLIT flag.

Yes, and that was bad because it created a hidden special case. The MARKER_SPLIT flag makes it clear at the top level that a Dividend transaction will create a marker split, that a buy will generate an AMOUNT_DEBIT and a VALUE_DEBIT while a stock split will generate only an AMOUNT_DEBIT.

> Moreover, FieldMask::DISABLED was intended to fully denote fields which should skip the page i.e. in my original design, the dividend page should skip both stock_amt and stock_val. In this branch, the "Stock Split" type will define stock_amt as FieldMask::DISABLED | FieldMask::AMOUNT_DEBIT | FieldMask::INPUT_NEW_BALANCE which is identical to FieldMask::AMOUNT_DEBIT | FieldMask::INPUT_NEW_BALANCE, i.e. it seems odd.

Correct, FieldMask::DISABLED can't be meaningfully combined with anything else because it sets no bit. That makes it useless to the program, but it's still useful as a placeholder in the FieldMask declarations. I agree that it's not helpful on the stock entry because there's always something else. I'll remove it tomorrow.

christopherlam commented:
Ok one more modification: `gnc_account_sel_set_new_account_modal (GNC_ACCOUNT_SEL(m_selector), true);` so that the "New account" button, having created the new account, will select it in the GNCAccountSel.

jralls committed:
[stock-txn-asst] Remove superflouos FieldMask::DISABLED tags

From the stock entry of stock split and reverse split templates. Since
there are other flags on the line the DISABLED flag isn't useful.

jralls commented:
> Ok one more modification: 

Done.
